### PR TITLE
Initial rough draft of How to Fetch in the browser

### PR DESCRIPTION
**Pull request by twalker**
### Summary

A rather rough draft to share for editorial suggestions and critique. 
It feels very amateurish and perhaps too brief, but I'm putting it out there. 

Please comment inline.

### What kind of change does this PR introduce?

- [x] Add a new feature/page

Related issue: 

### How to verify locally

1. `npm start`
1. http://localhost:8080/Tech-Writing-Project/browser-fetching/

### CODE REVIEW

File: _includes/layouts/base.njk [hunk 1]
```diff
@@ -26,18 +26,6 @@
     </header>
 
     <main{% if templateClass %} class="{{ templateClass }}"{% endif %}>
-
```
Comment (twalker) [inline _includes/layouts/base.njk]: Unrelated change:
This yellow notice was distracting and unpleasant to look at. I removed it.

File: browser-fetching/index.md [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+
+#### Image from the response:
+<div style="width: 400px;">
+  <img id="dog" style="width: 100%;" alt="Random dog image" src="https://via.placeholder.com/400" />
+</div>
+
+Below is more complex example that uses multiple options to POST a request to a weather API:
+
+```
+TODO
```
Comment (twalker) [inline browser-fetching/index.md]: @cmneal26 ,
Have you had a chance to identify an API to use in this more complex example?

File: _includes/layouts/base.njk (hunk 1 above)
Comment (ghost) [reply _includes/layouts/base.njk]: yes, it was aweful!!!!!!!!!

File: browser-fetching/index.md (hunk 2 above)
Comment (ghost) [reply browser-fetching/index.md]: Tim is right. Maybe, another team, you have to worry about this. But we got your back. If you'd like to do something with calling an api for an environmental api, and we can def. do that. It's all up to you. Anything you can think of, we can prob. make some headway there.

File: browser-fetching/index.md [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+```
+
+## Summary
+
+The previous examples only hint at the capabilities of the Fetch API. Read the ["Using the Fetch API"](https://developer.mozilla.org/en-US/docs/Web/API/Fetch_API/Using_Fetch) page on the Mozilla Developer Network (MDN) for more comprehensive examples and detail on the Fetch API. Fetch isn't limited to web browsers—it's also available in [Node.js v18](https://nodejs.org/en/blog/announcements/v18-release-announce/#fetch-experimental) and can be used on the server.
+
+Fetch is an essential API for any dynamic webpage that reacts to user interactivity. If you weren't fetching before, you sure are fetching now.
```
Comment (ghost) [inline browser-fetching/index.md:79]: I like the snark. It fits the 'playful' aspect we cultivate on some of our site pages, while respecting the "simple and clear" principle that guides our education content.

File: browser-fetching/index.md [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+  _For example, fetching current weather conditions for the given location_
+* **Partially updating the page** in response to user interactivity
+  _For example, fetching and displaying the current stock price for a user-provided stock index_
+* **Submitting forms** and displaying a success or error messages without needing to reload an entire page.
+  _For example, Adding an item to a shopping cart_
+
+Fetch's interface only requires a single argument—a `url` address to send the request to. The second argument is an object for specifying various options.
+
+```javascript
+  fetch(url, options) // return a Promise
```
Comment (ghost) [inline browser-fetching/index.md]: @cmneal26 

Tim, help me! What is a promise

File: browser-fetching/index.md [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+It's useful to web developers in scenarios such as:
+
+* **Displaying dynamic data** that changes over time
+  _For example, fetching current weather conditions for the given location_
+* **Partially updating the page** in response to user interactivity
+  _For example, fetching and displaying the current stock price for a user-provided stock index_
+* **Submitting forms** and displaying a success or error messages without needing to reload an entire page.
+  _For example, Adding an item to a shopping cart_
+
+Fetch's interface only requires a single argument—a `url` address to send the request to. The second argument is an object for specifying various options.
```
Comment (ghost) [inline browser-fetching/index.md]: love this line! i like the way it clearly expresses the idea.

File: browser-fetching/index.md [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+    console.log(result)
+    document.querySelector('#dog').src = result.message
+    document.querySelector('#dog-results').innerHTML = JSON.stringify(result, null, 2)
+  })
+  // Log any errors to the developer console
+  .catch(error => console.error(error))
+```
+
+<script>
+const fetchDog = () => fetch('https://dog.ceo/api/breeds/image/random')
```
Comment (ghost) [inline browser-fetching/index.md:55]: Here's the context. I'm someone from an under-priveleged background who sees this line of code. It strikes me as very intimidating, even though it's not really. It's an issue of perception.

How can we mitigate that intimidation factor here for the code block, insofar as it's possible in the ocntext? For example stuff like images and bright colors suggest feelings of security sometimes. That might be one way to go.

File: browser-fetching/index.md [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+
+#### JSON from the response:
+<pre class="language-javascript" id="dog-results">Nothing fetched yet.</pre>
+
+#### Image from the response:
+<div style="width: 400px;">
+  <img id="dog" style="width: 100%;" alt="Random dog image" src="https://via.placeholder.com/400" />
+</div>
+
+Below is more complex example that uses multiple options to `POST` a request to a weather API:
```
Comment (ghost) [inline browser-fetching/index.md]: Since we have two examples in this learning module, we can focus on the first one about the dog stuff first. We can make it shiny, and leave a placeholder for the second. If we get to it, fine. If not, that works, as well. We can then focus on getting the explanation for the first example right.

File: browser-fetching/index.md [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+
+```
+TODO
+```
+
+## Summary
+
+The previous examples only hint at the capabilities of the Fetch API. Read the ["Using the Fetch API"](https://developer.mozilla.org/en-US/docs/Web/API/Fetch_API/Using_Fetch) page on the Mozilla Developer Network (MDN) for more comprehensive examples and detail on the Fetch API. Fetch isn't limited to web browsers—it's also available in [Node.js v18](https://nodejs.org/en/blog/announcements/v18-release-announce/#fetch-experimental) and can be used on the server.
```
Comment (ghost) [inline browser-fetching/index.md]: We need to keep in mind context. Let's just stipulate that our outreach program has both adults and younger learners, who are unfamiliar with tech. So, we have to somehow balance this goal of catering to these learners with different inclinations with our writing/visual effects.

I thought of a useful idea to help us handle this situation, and appeal to a broader audience.

**THE KID BOX**

Think of this, it's like a visual box, with some additional information. It's shiny and colorful, and the idea is this.

A kid comes to site. They are on the learning module page. Although they're following the same module as the adult learner, there are these boxes that say something like "Kids only!" This will do a few things for us. It will allow us to add optional information that people can look at if they want (and it may be helpful to some adults, too, we don't know, so we cast a wide net). These boxes help to augment the kid's understanding, and now we have a flexible mechanism that allows us to hopefully tailor our explanation to target audiences that are distinct in important ways, in terms of how they approach learning.

Comment (ghost) [reply browser-fetching/index.md]: This also allows us to not force our simple and clear content to be juvenile and not appeal to kids. if kids need more info, put it in the kid box. But the kid box is also more versatile than that. It can be used not only to provide information for kids, but to do something else with language or visual affects. What? Sometimes a piece of language can be used to entertain, for example, and that can be useful.

Comment (ghost) [reply browser-fetching/index.md]: This also helps with our branding. Parents love kids content. They'd be more receptive to our site if they see it as something for kids. And, this in turn, plays into our long term gain of securing more donations and gaining more exposure so that we can get money, and pursue the Free Free API Education company's goal to the best of our ability.

File: browser-fetching/index.md [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+    document.querySelector('#dog').src = result.message
+    document.querySelector('#dog-results').innerHTML = JSON.stringify(result, null, 2)
+  })
+  // Log any errors to the developer console
+  .catch(error => console.error(error))
+```
+
+<script>
+const fetchDog = () => fetch('https://dog.ceo/api/breeds/image/random')
+  .then(response => response.json())
```
Comment (ghost) [inline browser-fetching/index.md:56]: This is something I'm not sure about.

Should we include inline comments, like

```javascript

const iFetchedADuck = {name: "I am a duck"} // some comment that explains this; more verbose than your usual

```

Comment (ghost) [reply browser-fetching/index.md:56]: Here is a thought. So, what would be most useful to our users eventually is to be able to use real code comments to make sense of code. If we think of this as an iterative process ,and don't even tell them, and use comments like real comments, then we get the added benefit of teaching them something "under the hood" -- something they might not realize for awhile. But remember, our non-profit is playing the long game.

File: browser-fetching/index.md [hunk 10]
```diff
@@ -0,0 +1,80 @@
...
+  .then(result => {
+    console.log(result)
+    document.querySelector('#dog').src = result.message
```
Comment (ghost) [inline browser-fetching/index.md:59]: I like that the naming is simple and semantic (the names themselves go some way toward helping you understand what the thing named does).

However, I think we should simplify this a little, keeping in mind our target audience.

How do we do that? One way.
```javascript
const dogBox = document.querySelector('#dog') // reference to visual box (or whatever)
dogBox.src = result.message // assign the message to the dogBox (something like this)
```

File: browser-fetching/index.md [hunk 11]
```diff
@@ -0,0 +1,80 @@
...
+    document.querySelector('#dog-results').innerHTML = JSON.stringify(result, null, 2)
+  })
+  .catch(error => console.error(error))
+</script>
+<button onclick="fetchDog()">Call the `fetchDog` function</button>
+
+#### JSON from the response:
+<pre class="language-javascript" id="dog-results">Nothing fetched yet.</pre>
+
+#### Image from the response:
```
Comment (ghost) [inline browser-fetching/index.md:69]: Each sub-section can have a "Kid Box", we either put something entertaining, or something that will help augment their understanding in a way that might help a kid continue with the lesson.

File: browser-fetching/index.md (hunk 8 above)
Comment (ghost) [inline browser-fetching/index.md]: I think we want to emphasize more the magic of APIs. This would be a good place for a little more persuasive writing. We're always transparent, but in this case we try to persuade them a little more that this is something worth learning.

Comment (ghost) [reply browser-fetching/index.md]: Actually, I think we want two kinds of supplementary information boxes. Perspective and Kids.

We can think about the exact styling for these later. What we want to do first is evaluate them from a structural and educational perspective.

Use the Perspective BOX (name pending) to provide perspective wherever it is warranted. Now we have a standard mechanisms that everyone understands how to use. And we can keep the styles centralized for this in one place, which will make it super handy to just focus on content.

Comment (ghost) [reply browser-fetching/index.md]: And, for example, if the function of a piece of writing is to persuade. Sometimes that will be the case. You don't necessarily have to write it, if you feel more comfortable with someone else doing that kind of writing. For example, you could be like "hey, team member X, can I get a persuasive paragraph on this topic? I'm pretty good with simple and clear style but that might not mean im the best persuasive writer, and that's okay. Use team's resources.

File: browser-fetching/index.md (hunk 11 above)
Comment (ghost) [reply browser-fetching/index.md:69]: Maybe it turns out that having them in every sub-section is overkill. But we wont know until we implement it and stare at the visual output.

File: browser-fetching/index.md (hunk 3 above)
Comment (ghost) [inline browser-fetching/index.md:79]: Although we likely won't have anymore education modules on the site for this class, we should still add something like a button like "next module".

Oh, and what do you think of naming the navigation title for this page, "Learn". It takes them immediately to your content module. However, although it would be final for our class project, if it were real, we'd "Learn" would go to some sort of intermediary page, with maybe some submodule nav bar. for example, down the road (not this class),

File: browser-fetching/index.md (hunk 10 above)
Comment (ghost) [reply browser-fetching/index.md:59]: I don't know if this is simpler to our target audience, even though it strikes me that way now. This might be something to look into a usability test for.

File: browser-fetching/index.md (hunk 3 above)
Comment (ghost) [reply browser-fetching/index.md:79]: Oh, we will possibly have that environmental tutorial, or something like that. My thought here is that although it would be better if all of the educational content was gotten at through the "Learn" top level nav link, for the context of this project, I would be fine having two high level nav items, the fetching one, and the environment science one. 

But, we recognize that one step in the future that would be good from the perspective of site design would be to move all the education modules into the top the 'Learn' area.

File: browser-fetching/index.md [hunk 12]
```diff
@@ -0,0 +1,80 @@
...
+eleventyNavigation:
+  key: Fetch
+  order: 4
+---
+
+# Aren't you fetching?!
+
+_I'm uncomfortable, you're probably uncomfortable, let's move on to fetching in the web browser_.
+
+All modern web browsers supply a Fetch API for making a HTTP requests to a web server, and handling the data returned from the server. The Fetch API provides a programming interface to asynchronously exchange data with a web server without needing to load another web page.
```
Comment (cmneal26) [inline browser-fetching/index.md]: small grammar error here - "a HTTP requests"

File: browser-fetching/index.md (hunk 5 above)
Comment (cmneal26) [inline browser-fetching/index.md]: For newbies it might be good to define "argument" and "object"

**Review by cmneal26 — APPROVED** [review]

**Comment by ghost** [comment]
@twalker If you don't mind, once you make a bit of progress on this, feel free to merge it, add as 'Fetching' (whatever name makes sense.) top-level nav page. 

I think it will be easier to provide you with comments if the partially completed work is on the publically available page, because we can see the visual representation without having to spin up local.

File: browser-fetching/index.md (hunk 6 above)
Comment (ghost) [reply browser-fetching/index.md:55]: @twalker I thought of an interesting idea that could help us address this. You know those pastoral paintings. The ones that show, like, "a farm or a field or some rolling hills." and they're kind of soothing. Maybe we introduce for our learning modules a light visual theming element. What do I mean? Imagine if, for example, we had leaves, or some naturey thing that can sooth in key places as ornaments -- but not distract from the content.

Comment (twalker) [reply browser-fetching/index.md:55]: That's a good idea. I think I'm misunderstood that the audience is no longer an environmental scientist. I'll consider a rewrite for the new target audience.

File: browser-fetching/index.md (hunk 9 above)
Comment (twalker) [reply browser-fetching/index.md:56]: Indeed, comments are great for explanations. I've updated the example `fetchDog` function to explain a bit more what each line does. Thanks.

File: browser-fetching/index.md (hunk 12 above)
Comment (twalker) [reply browser-fetching/index.md]: Thank you.

**Review by ghost — APPROVED** [review]

File: browser-fetching/index.md (hunk 5 above)
Comment (twalker) [reply browser-fetching/index.md]: It's a challenge to decide which terms to explain and which can be assumed knowledge. Instead of provide inline definition of `argmument`, `object`, and `Promise`, I linked the terms to reference articles. It's fantastic to know what terms warrant some more information. 👍🏼
